csharp/src/iceboxnet/ServiceManager.cs

```
@@ -441,295 +398,479 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
+                        {
+                            if (!startTask.IsValueCreated)
+                            {
+                                waitTask = startTask;
+                            }
+                            else if (startTask.Value.IsFaulted || startTask.Value.IsCanceled)
+                            {
+                                // Already stopped, no need to do anything.
+                                _startTask = null;
+                                _stopTask = new(Task.CompletedTask);
+                                performTask = _stopTask;
+                            }
+                            else if (!startTask.Value.IsCompletedSuccessfully)
+                            {
+                                waitTask = startTask; // we need to wait until this task is complete
+                            }
+                            // else, startTask completed successfully and we replace it with a stopTask below.
+                        }
+                        else
+                        {
+                            Debug.Assert(false);
+                            performTask = new(Task.CompletedTask);
+                        }
+
+                        if (waitTask == null && performTask == null)
+                        {
+                            _startTask = null;
+
+                            // The async/await ensures we get a Task (i.e. successful lazy initialization) even if
+                            // StopAsync throws an exception synchronously.
+                            // Note that StopAsync is not executed immediately with _mutex locked.
+                            _stopTask = new(async () => await _service.StopAsync());
+                            performTask = _stopTask;
+                        }
+                    }
+
+                    if (waitTask != null)
+                    {
+                        await waitTask.Value;
```

same here, should we ignore the exceptions and return a completed task if it fails?

---

Fixed.